### DISCUSSION TITLE
Filenames are not correct encoded on expanding old archives build with version 0.86 (i.e.)

### DISCUSSION THREAD

**Discussion by freica** [icsharpcode/SharpZipLib#686] [Q&A]
### Steps to reproduce
1. build an zip-archive using version 0.86 with filenames conaining special chars i.e äöüÄÖÜß...
2. expand this zip-archive using current version (1.3.3)
3. filenames with special chars are not correct encoded
4. archives built with the current version are well expaned

**Comments:**

- 7zip i.e. can expand archives build with version 0.86 as well as archives built with version 1.3.3
- On the compress side we used only default encoding and nothing else, assuming UTF  will be used and all is done well.

**Questions:**
1. What's to do to get always well encoded filenames on expanding an archive undependent of the used zip-version.?

**Comment by piksel** [comment]
UTF-8 is used if the Unicode bit is set for the entries, so set it using:
```cs
zipEntry.IsUnicodeText = true;
```
Otherwise the encoding is set to the "system default" encoding, which may mean different things depending on OS, language settings, framework and sharpziplib version.

**Reply by freica** [reply]
I tried it out, but it does not the job, perhaps there is a misunderstanding?
The original filename is: Test_äöüÄÖÜß'.txt, but the expanded looks like the attachment independant of IsUnicodeText is set to true or not.

```
      ZipFile file = null;
      FileStream fs = File.OpenRead(FileZipPath);
      file = new ZipFile(fs);

      foreach (ZipEntry zipEntry in file)
      {
        if (!zipEntry.IsFile)
        {
          // Ignore directories
          continue;
        }

        zipEntry.IsUnicodeText = true;

        String entryFileName = zipEntry.Name;
      }

```

![ExpandedFileName](https://user-images.githubusercontent.com/25507027/142161839-a42cc49c-67d1-49f5-ac75-9cc26a456b5b.png)

**Reply by piksel** [reply]
No, I mean when you create the archive. This is what tells the unarchiver what encoding the file names where written in and is the "correct" way do it as per the zip specification. If that flag hasn't been set in the archive, it *should* use a specific IBM/DOS encoding, but this is largely ignored by most zip programs who each do their own thing instead (generally they use the OS default encoding, but that can vary).

If you want to force unicode when extracting, you can set it using a static property:
```cs
ICSharpCode.SharpZipLib.Zip.ZipStrings.UseUnicode = true;
```

**Reply by freica** [reply]
At the create-time we have not set the encoding (not in the past when using V0.8.6 and not now when using V1.3.3)
At the extracting-time we set also nothing about encoding.
Now after extracting always with V1.3.3 we see the following result (Win 10, German):

- files which were compressed with V1.3.3 have a good encoding
- files which were compressed with V0.8.6 have a bad encoding (using ICSharpCode.SharpZipLib.Zip.ZipStrings.UseUnicode = true does not help because UniCode-Encoding was not used at compress-time)

Before switching extracting to V1.3.3, the V0.8.6 archives were well encoded
So what should we do to get old archives also well encoded using V1.3.3 to extract?

**Reply by freica** [reply]
Can we please continue this issue?

**Reply by piksel** [reply]
You can set the codepage using the `ICSharpCode.SharpZipLib.Zip.ZipStrings.CodePage` property. I don't know what code page you used to create them with, but `1251` is a good guess (ISO-8859-1 / Latin1 / "Western European").

**Reply by freica** [reply]
At create time we also don't set the codepage, so over all version the default settings are used. 
So please tell me what is the default codepage used at compress time by SharpZipLib and what's the difference on this setting over the SharpZipLib versions resulting in the above described behaviour.

Is there a chance to identify the used codepage on reading a zip-file with SharpZipLib?

**Reply by piksel** [reply]
> Is there a chance to identify the used codepage on reading a zip-file with SharpZipLib?

No, it used the "system default" encoding, which varies between versions of OSes and frameworks (.NET Framework vs .NET Core etc). And that is not stored inside the zip. This is the "wild west" situation that winzip et al created, but "our" solution for looking up the encoding doesn't work the same way in .NET Core/5+.
For .NET framework it's determined the equivalent of this setting:
![image](https://user-images.githubusercontent.com/807383/144206475-9d3979df-ac2e-49b1-bfac-dad58e859476.png)
(at least, as far as I understand the documentation, I have not tested this).
If you send me a sample file I can probably figure it out...

Also note that if you are running on anything newer than .NET 4.8 you may need to call:
```cs
Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
```
to be able to use the codepage.
<details>
<summary>Encoding tests</summary>

```
❯ dotnet run -f net6.0
Framework: .NET 6.0.0-rtm.21522.10
Default encoding: Unicode (UTF-8), Codepage: 65001, WinCodepage: 1200

Attempting to retrieve codepage 1252...
Failed to get code page: System.NotSupportedException: No data is available for encoding 1252. For information on defining a custom encoding, see the documentation for the Encoding.RegisterProvider method.
   at System.Text.Encoding.GetEncoding(Int32 codepage)
   at Program.<<Main>$>g__tryGetCP|0_0(Int32 cp) in Program.cs:line 11

Registering additional codepage providers...
Attempting to retrieve codepage 1252...
Successfully retrieved codepage 1252!
```

```
❯ dotnet run -f net48
Framework: .NET Framework 4.8.4420.0
Default encoding: Western European (Windows), Codepage: 1252, WinCodepage: 1252

Attempting to retrieve codepage 1252...
Successfully retrieved codepage 1252!
```

</details>

**Reply by freica** [reply]
But I figured out it depends also from the default settings of SharpZipLib.
(For both tests I am using .NET 4.7.2)

Using the old version 0.86 :
ZipConstants.DefaultCodePage is initialized by SharpZipLib with **850**
[MyOutputFileV086.zip](https://github.com/icsharpcode/SharpZipLib/files/7632903/MyOutputFileV086.zip)
Here 7zip shows a correct encoding
![grafik](https://user-images.githubusercontent.com/25507027/144215926-809bba92-a100-458c-815f-0482f2684a7a.png)

Using the current version 1.33 :
ZipConstants.DefaultCodePage is initialized by SharpZipLib with **65001**, but the dump of IsUnicodeText shows **false**
[MyOutputFileV133.zip](https://github.com/icsharpcode/SharpZipLib/files/7632905/MyOutputFileV133.zip)
But here 7zip shows not the correct encoding:
![grafik](https://user-images.githubusercontent.com/25507027/144215779-bbeeefa5-98a4-46ca-b02d-8738b290af51.png)

This is the filename I compress in both test:
![grafik](https://user-images.githubusercontent.com/25507027/144216709-430176c4-7477-48f1-83f4-31bbb94a5258.png)

This are my constant os settings for both tests:
![grafik](https://user-images.githubusercontent.com/25507027/144216187-1445abfa-ce30-421c-9349-bdc5e18a2950.png)

**Reply by piksel** [reply]
The `ZipConstants.DefaultCodePage` (and `ZipStrings.CodePage`) is by default set to `Automatic`, which means "use UTF-8 if the unicode bit is set, otherwise use `ZipStrings.SystemDefaultCodePage`.
Reading that value returns the UTF-8 code page since it's the default when creating archives (yes, the way the encodings were implemented were a mess).
https://github.com/icsharpcode/SharpZipLib/blob/v1.3.3/src/ICSharpCode.SharpZipLib/Zip/ZipStrings.cs#L47

**Reply by piksel** [reply]
Your file has been encoded using CP-1252:
```
> dotnet run -f net48
Framework: .NET Framework 4.8.4420.0
Encoded bytes, default:     54, 65, 73, 74, 5f, e4, f6, fc, c4, d6, dc, df, 27, 2e, 74, 78, 74
Encoded bytes, unicode bit: 54, 65, 73, 74, 5f, c3, a4, c3, b6, c3, bc, c3, 84, c3, 96, c3, 9c, c3, 9f, 27, 2e, 74, 78, 74
Encoded bytes, cp850:       54, 65, 73, 74, 5f, 84, 94, 81, 8e, 99, 9a, e1, 27, 2e, 74, 78, 74
Encoded bytes, cp437:       54, 65, 73, 74, 5f, 84, 94, 81, 8e, 99, 9a, e1, 27, 2e, 74, 78, 74
Encoded bytes, cp1252:      54, 65, 73, 74, 5f, e4, f6, fc, c4, d6, dc, df, 27, 2e, 74, 78, 74
```
(it matches the filename bytes in the report: https://archivediagstore.blob.core.windows.net/reports/VxT03lH_zUqpLzs5FlfpeQ.html)

It looks like 0.86 used `Thread.CurrentThread.CurrentCulture.TextInfo.OEMCodePage` instead as the default, which is probably why you were seeing cp850 (`Western European (DOS)`).

**Reply by freica** [reply]
Ok, this are the properties I get using V1.3.3, so the system-codepage seems to be 1252 here.
![grafik](https://user-images.githubusercontent.com/25507027/144244725-714f0ef9-e2bb-4494-a0e7-828c38119951.png)

But using the old version V0.8.6 there is no ZipStrings class, but the system settings must be the same. On the other hand the used encoding is different. So what's the matter ?

**Reply by piksel** [reply]
> It looks like 0.86 used Thread.CurrentThread.CurrentCulture.TextInfo.OEMCodePage instead as the default, which is probably why you were seeing cp850 (Western European (DOS)).

**Reply by freica** [reply]
Ok, seems possible...
![grafik](https://user-images.githubusercontent.com/25507027/144256072-05fa053f-5fe2-4065-af83-bb1a3604c84c.png)

**Reply by freica** [reply]
So there is only one thing from this morning not solved. Using V1.3.3 and default settings for example 7zip does show a bad encoding of the filename. 
A further test with 
`ZipStrings.UseUnicode = true;`
results in a good encoding in 7zip.

On the other hand the implemented encoding is realy a mess, I had a look at the code.
And here one thing in my mind could be a bug.

```
static ZipStrings()
		{
			try
			{
				var platformCodepage = Encoding.GetEncoding(0).CodePage;
				SystemDefaultCodePage = (platformCodepage == 1 || platformCodepage == 2 || platformCodepage == 3 || platformCodepage == 42) ? FallbackCodePage : platformCodepage;
			}
			catch
			{
				SystemDefaultCodePage = FallbackCodePage;
			}
		}

private static int codePage = AutomaticCodePage;

private const int AutomaticCodePage = -1;

public static int CodePage
		{
			get
			{
				return codePage == AutomaticCodePage? Encoding.UTF8.CodePage:codePage;
			}
			set
			{
				if ((value < 0) || (value > 65535) ||
					(value == 1) || (value == 2) || (value == 3) || (value == 42))
				{
					throw new ArgumentOutOfRangeException(nameof(value));
				}

				codePage = value;
			}
		}

```
If no settings are made from outside, `codePage` and `CodePage` is not the same!
`codePage` results in -1 and `CodePage` in Encoding.UTF8.CodePage (=65001)!
A further result of this is that `UseUnicode` in this case results in `false`!

(By the way, I don't no how you can compile `return codePage == AutomaticCodePage? Encoding.UTF8.CodePage:codePage` 
Here my Compiler runs in an error, because `AutomaticCodePage` is not a `bool`)

Please have a look on it ;-)

**Reply by piksel** [reply]
`codePage == AutomaticCodePage` is a bool.

```cs
return codePage == AutomaticCodePage ? Encoding.UTF8.CodePage : codePage;

// is the same as:

if (codePage == AutomaticCodePage) {
	return Encoding.UTF8.CodePage;
} else {
	return codePage;
}
```

The `CodePage` getter and setter does not map 1:1, this is because the same static settings are used for both creating and reading zip files, so there is no easy way to answer "what is the encoding/codepage being used", as the encoding depends on the entry headers.
The internal `codePage` is either a forced codepage specified by the consumer, or it's `AutomaticCodePage`.

`UseUnicode` forces the use of `UTF-8` regardless of what the entry headers indicate, hence why it returns false if you have not specifically set it.

**Reply by freica** [reply]
Ok, I understand
`return codePage == AutomaticCodePage ? Encoding.UTF8.CodePage : codePage;` Here I was to fast with my eyes,

But if I set UseUnicode=true, it seems that I get an other encoding (UTF-8) which 7zip shows correct.
So should this say that 7zip could identify on my system archives encoded with codepage 65001(UTF-8) and 850 and perhaps other ones, but not 1252?

If the difference between `codePage` and `CodePage` is not a bug, so at least it's a questionable implementation that encourages this mess.

Have a nice evening!
